ICU-9562 Update language tag mapping per the latest IANA registry

jungshik opened the pull request:
uloc_forLanguageTag has a few mapping tables to map grandfathered
language tags and deprecated language subtags to their preferred or
modern values.

Update them based on the latest version of the IANA language subtag registry. [1]

Five grandfathered tags without a preferred value are still mapped to
what ICU has mapped them to for the backward compatibility until the
wisdom of continuing to do so is reviewed.

In addition, map redundant language tags to their preferred values
regardless of whether they're followed by other subtags or not. (e.g.
zh-yue vs zh-yue-u-co-pinyin) .

Similary, ja-latn-hepburn-heploc is mapped to ja-latn-alaic97 (the
variant subtag 'hepburn-helploc' with the prefix 'ja-latn' has the
preferred value, 'alaic97') .

Update the mapping for deprecated language subtags (e.g. 'jw' to
'jv' and a bunch of 3-letter language codes).

Add a new table for deprecated region subtags to map them to their
modern values. (e.g. 'DD' to 'DE').

Add a new test case for deprecated language and region mapping and
a few more cases for updated grandfathered and redundant tag mapping.

[1]
https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry

<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [x] Issue filed at https://unicode-org.atlassian.net :  ICU-______
- [x] Update PR title to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

jungshik commented:
@yumaoka ,  can you take a look?  Thanks !

@roubert : FYI

jungshik commented:
I'll make a separate PR for Java.

jungshik commented:
Hold on. Rebasing made a new test fail. Let me fix that.

jungshik commented:
it turned out that git misapplied a part of my change at a wrong place during rebase. It's fixed and all the tests should pass now.

jefgen commented on icu4c/source/common/uloc_tag.cpp.orig:
```diff
@@ -0,0 +1,2534 @@
+// © 2016 and later: Unicode, Inc. and others.
```
Should this file be included?

jungshik replied:
oops.  No. I'll delete them, too. Hmm.. I don't know what I was doing.. 

Thank you for catching it.

jefgen commented on icu4c/source/common/uloc_tag.cpp.rej:
```diff
@@ -0,0 +1,22 @@
+--- source/common/uloc_tag.cpp
```
nit: Looks like maybe this got added by accident?

jungshik replied:
oops. thank you for the catch.

roubert commented:
This test failure needs some further investigation:

`FAIL: en-GB-oed; got "en_GB_OXENDICT"; expected "en_GB@x=oed"`

The strings on the right hand side, `"en_GB_OXENDICT"` and `"en_GB@x=oed"` are the ICU internal representation (our legacy locale ID strings) and the updated mapping of the BCP-47 grandfathered language tags has caused this to change.

Is that right? Should this change? Is `"en_GB_OXENDICT"` really a valid locale ID? Might there be code out there expecting `"en_GB@x=oed"` that now will fail?

If this all is as it should be, then the expected string would just need to be updated in `TestForLanguageTag()` in `icu4c/source/test/intltest/loctest.cpp`.

jungshik commented:
The test was written with the old data in mind. I just updated the expectation. The expected value should be en_GB_OXENDICT  (en-GB-oxendict is the preferred value for en-GB-oed per the latest IANA language subtag registry. ).  Before this PR, en-GB-oed was converted to en-GB-x-oed leading to en_GB@x=oed after uloc_forLanguageTag.

jungshik commented:
> "en_GB_OXENDICT"

I believe that that's the way BCP47's variant subtag is handled by uloc_forLanguageTag.

roubert commented:
This question remains unanswered: Is the legacy ICU locale ID `"en_GB_OXENDICT"` both valid and the correct one to use for Oxford English?

markusicu commented:
Since 2015, en-GB-oxendict is the proper replacement for en-GB-oed.

See https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
```
Type: grandfathered
Tag: en-GB-oed
Description: English, Oxford English Dictionary spelling
Added: 2003-07-09
Deprecated: 2015-04-17
Preferred-Value: en-GB-oxendict
```

jungshik commented:
> This question remains unanswered: Is the legacy ICU locale ID "en_GB_OXENDICT" both valid and the correct one to use for Oxford English?

The answer is positive. See how 'de-1901', 'de-Latn-1901' or  'de-DE-1901' is handled by uloc_forLanguageTag.  '1901' is a variant subtag.
 
@yumaoka can chime in. :-)

jungshik commented:
https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry has this : 

```
%%
Type: variant
Subtag: oxendict
Description: Oxford English Dictionary spelling
Added: 2015-04-17
Prefix: en
%%
```

jungshik commented:
Another exhibit:  
```uloc_toLanguageTag( uloc_forLanguageTag ("en-GB-oed"))``` gives back 'en-GB-oxendict'.

roubert commented:
I took a look at the CLDR data and the ICU data built from it, and what I found was that OXENDICT is registered there as a variant which means that `"en_GB_OXENDICT"` is the correct locale ID to use.

A quick test with `icu::Locale::getDisplayName()` shows that this indeed is the locale ID that ICU itself understands to be Oxford English:

```
Locale ID    : "en_GB_OXENDICT"
Display Name : "English (United Kingdom, Oxford English Dictionary spelling)"

Locale ID    : "en_GB@x=oed"
Display Name : "English (United Kingdom, Private-Use=oed)"
```

This means that ICU has been doing this wrong, converting the BCP-47 language tag `"en-GB-oed"` into an internal representation that it doesn't understand itself, for the past 8 years. I had not expected that.

roubert commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -1938,6 +2125,37 @@ ultag_parse(const char* tag, int32_t tagLen, int32_t* parsedLen, UErrorCode* sta
 
+    size_t parsedLenDelta = 0;
+    if (grandfatheredLen == 0) {
+        for (i = 0; REDUNDANT[i] != nullptr; i += 2) {
```
Instead of depending on this list to be null-terminated, I suggest that you use `UPRV_LENGTHOF()` for the loop condition, like you already do for the other lists.

jungshik replied:
Done. Thanks

roubert commented on icu4c/source/test/cintltst/cloctst.c:
```diff
@@ -6156,6 +6167,44 @@ static void TestInvalidLanguageTag(void) {
+    char canonical[256];
+    int32_t i;
+    UErrorCode status;
+    for (i = 0; langtag_to_canonical[i].input != NULL; i++) {
```
Also here, iterating over the allocated size is both simpler and safer.

jungshik replied:
Done

roubert commented on icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/ULocaleTest.java:
```diff
@@ -4110,6 +4110,7 @@ public void TestForLanguageTag() {
                 {"art-lojban",          "jbo",                  NOERROR},
                 {"zh-hakka",            "hak",                  NOERROR},
                 {"zh-cmn-CH",           "cmn_CH",               NOERROR},
+                {"zh-cmn-CH-u-co-pinyin", "cmn_CH@collation=pinyin", NOERROR},
```
This random Java change looks as if it has been included in this pull request by mistake.

jungshik replied:
Well, I included it on purpose to sync C test and Java test as far as both pass, but I can drop it from this PR and lump it together with an upcoming Java PR.

jungshik commented:
BTW, this PR does not deal with handling of two variant subtags for 'hy' that were deprecated this year with preferred values.  I've just discovered it while looking at one of tests in Ecma 262 test suite v8 fails. It's a bit tricky to handle correctly ("arevela" has to be dropped when lang subtag is hy. "arevmda" has to be dropped and 'hy' has to be replaced by 'hyw' when the lang subtag is hy) , but there might be a way (the way preeuro is handled may be utilized?) . 

```
%%
Type: variant
Subtag: arevela
Description: Eastern Armenian
Added: 2006-09-18
Deprecated: 2018-03-24
Preferred-Value: hy
Prefix: hy
%%
Type: variant
Subtag: arevmda
Description: Western Armenian
Added: 2006-09-18
Deprecated: 2018-03-24
Preferred-Value: hyw
Prefix: hy
%%
```

yumaoka commented:
>@yumaoka , do you have any concern? If so, let me know. Otherwise, could you merge?

No. Looks fine to me. As long as you have a plan for ICU4J porting soon (after ICU 63), please go ahead and merge this changes to ICU4C (for ICU 63).
Thanks.

jungshik commented:
Thanks. I don't have a privilege to press merge button. Could you (or @roubert or @jefgen )  press the button?  Thanks :-)
